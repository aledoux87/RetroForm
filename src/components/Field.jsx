import React from 'react';
import { string, oneOf, oneOfType, number, func } from 'prop-types';

function Field({ id, name, type, label, placeholder, defaultValue, onChange }) {
  return (
    <div className="retro-field">
      {label && <label htmlFor={id}>{label}</label>}
      <input
        className="retro-input"
        id={id}
        name={name || id}
        type={type}
        placeholder={placeholder}
        defaultValue={defaultValue}
        onChange={onChange}
      />
    </div>
  );
}

Field.defaultProps = {
  type: 'text'
};

Field.propTypes = {
  id: string.isRequired,
  name: string,
  type: oneOf(['text', 'number', 'email', 'password']),
  label: string,
  placeholder: string,
  defaultValue: oneOfType([string, number]),
  onChange: func.isRequired
};

export default Field;
