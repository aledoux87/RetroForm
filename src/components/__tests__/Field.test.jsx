import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Field from '../Field';

describe('<Field />', () => {
  it('Should Component Renders With Default Behaviour', () => {
    const { container } = render(<Field id="fieldId" onChange={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it('Should Component Renders With Props', () => {
    const { queryByText } = render(
      <Field
        id="fieldId"
        name="fieldName"
        type="text"
        label="FieldLabel"
        placeholder="DefaultField"
        defaultValue="Default"
        onChange={jest.fn()}
      />
    );
    expect(queryByText('FieldLabel')).toBeInTheDocument();
  });

  it("Should 'onChange' Function Has Been Called", () => {
    const onChangeMock = jest.fn();

    const { container } = render(<Field id="fieldId" onChange={onChangeMock} />);

    const field = container.querySelector('#fieldId');

    fireEvent.change(field, { target: { value: 'HelloWorld' } });

    expect(onChangeMock).toHaveBeenCalled();
  });
});