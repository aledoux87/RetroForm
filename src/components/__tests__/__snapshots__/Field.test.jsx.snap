// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field /> Should Component Renders With Default Behaviour 1`] = `
<div>
  <div
    class="retro-field"
  >
    <input
      class="retro-input"
      id="fieldId"
      name="fieldId"
      type="text"
      value=""
    />
  </div>
</div>
`;